### PR TITLE
Add max_spilling_size to object spilling config

### LINKED ISSUE #24667: [Core] Spilling performance regression in large-scale shuffle
### What happened + What you expected to happen

It seems #22750 introduced performance regression in large-scale shuffle. This is 1TB shuffle, 1GB partitions, running on 10x `d3.2xlarge` VMs. I tested version 1.12.0 and commit bdd3b9a0abdb96f70584092ae4d97f74769533b3 (one before the PR was merged) and the run time is 638 seconds. After the PR (commit 1c972d5d2d076ec069a81e81b5d6ba6dc34a92ce), the run time becomes 1014s.

Disk utilization plot before:

before-638s

After:
after-1014s

The spilled amount is about the same (1.8TB). cc @stephanie-wang 

### Versions / Dependencies

1c972d5d2d076ec069a81e81b5d6ba6dc34a92ce and master

### Reproduction script

https://github.com/franklsf95/raysort

### Issue Severity

_No response_

### PR DESCRIPTION
**Pull request by frank-lsf**
## Why are these changes needed?

The problem is documented in #24667. TLDR: When Ray is configured to spill to multiple HDD devices for max I/O, we need to limit the size of a single fused spill file. If the file is too large, then it will only be spilled to one HDD, leaving the rest of the HDDs idle, hurting load balancing and thus aggregate I/O throughput.

## Related issue number

Closes #24667

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

cc @stephanie-wang @scv119 @rkooo567

### CODE REVIEW

File: src/ray/raylet/local_object_manager.cc [hunk 1]
```diff
@@ -169,29 +169,35 @@ bool LocalObjectManager::IsSpillingInProgress() {
...
+      int64_t size = it->second->GetSize();
+      if (bytes_to_spill + size > max_bytes_to_spill) {
+        break;
+      }
+      bytes_to_spill += size;
       objects_to_spill.push_back(it->first);
+      count += 1;
     }
     it++;
-    counts += 1;
```
Comment (stephanie-wang) [inline src/ray/raylet/local_object_manager.cc:188]: Huh, was this a bug before?

File: src/ray/raylet/test/local_object_manager_test.cc [hunk 2]
```diff
@@ -33,6 +33,8 @@ namespace ray {
 
 namespace raylet {
 
+const uint64_t MAX_SPILLING_SIZE = 4L * 1024 * 1024 * 1024;
```
Comment (stephanie-wang) [inline src/ray/raylet/test/local_object_manager_test.cc:36]: Could you add a unit test for max spilling size?

**Review by stephanie-wang — APPROVED** [review]
Nice fix! Were you able to confirm that this works for the performance regression in shuffle?

Main comment is to please add a C++ unit test. Thanks!

File: src/ray/raylet/local_object_manager.cc (hunk 1 above)
Comment (franklsf95) [inline src/ray/raylet/local_object_manager.cc:188]: @rkooo567 Can you double check this change? The original code will increment the count regardless of whether the object is added to `objects_to_spill`; this change made it only increment when it is added to the objects list.

File: src/ray/raylet/test/local_object_manager_test.cc [hunk 3]
```diff
@@ -634,7 +646,7 @@ TEST_F(LocalObjectManagerTest, TestSpillUptoMaxFuseCount) {
     objects.push_back(std::move(object));
   }
   manager.PinObjectsAndWaitForFree(object_ids, std::move(objects), owner_address);
-  ASSERT_TRUE(manager.SpillObjectsOfSize(total_size));
+  ASSERT_TRUE(manager.SpillObjectsOfSize(total_size, MAX_SPILLING_SIZE));
```
Comment (franklsf95) [inline src/ray/raylet/test/local_object_manager_test.cc:649]: @rkooo567 Two questions: 1. why is this test in `LocalObjectManagerTest` but not `LocalObjectManagerFusedTest`? I thought the former sets max_fused_count to 1 whereas the latter sets `max_fused_count` to 15, which is what is tested here. 2. Why does line 657 say "we limit the max fused number of objects to be 10"? Is this intended to be 15?

File: src/ray/raylet/test/local_object_manager_test.cc [hunk 4]
```diff
@@ -660,6 +672,55 @@ TEST_F(LocalObjectManagerTest, TestSpillUptoMaxFuseCount) {
   }
 }
 
+TEST_F(LocalObjectManagerMaxSpillingTest, TestSpillUptoMaxSpillingSize) {
```
Comment (franklsf95) [inline src/ray/raylet/test/local_object_manager_test.cc:675]: @stephanie-wang Could you help me check if this test looks right?

**Comment by stephanie-wang** [comment]
It looks like there is a related test failure in the windows suite[1], test_object_spilling_3:

::test_pull_spilled_object[multi_node_object_spilling_config0] (pid=) [2022-07-08 00:24:47,500 C 12844 7564] (raylet.exe) local_object_manager.cc:178:  Check failed: min_bytes_to_spill <= max_bytes_to_spill
--

**Comment by franklsf95** [comment]
Sorry guys I messed up the git signoff feature, please ignore this.

**Pull request closed without merging**